### PR TITLE
Easier access and indexing to depletion time points

### LINKED ISSUE #1713: Feature: Easier indexing through ResultsList to find depletion step index
If a user wanted to perform a restart simulation not necessarily at the end point in time, there isn't a super simple way to do it from the API. Currently you would have to manually search through the `Result` instances and compare their time data one at a time. 

This isn't a monumental task, and I think we could easily provide a method that would allow users to find the index corresponding to a specific point in time. Supporting similar `time_units` like we do for the `Operator` (days, seconds, MWd/kgHM) would also be super nice. 

Looking at #1708, this would also be useful when exporting `material.xml` files either during the depletion run or after the fact. Either way, the user will need a way to link a depletion step to a point in calendar time for restarting or other analysis.

Other things to consider would be some tolerance on the search criteria. If I ask for the step where burnup was 12 MWd/kgHM, (for example), but the closest point is 12.1 MWd/kgHM, should that step be returned? Or error out?

### PR DESCRIPTION
**Pull request by drewejohnson**
This PR introduces two new `ResultsList` methods. 
- `ResultsList.get_times` returns the depletion schedule (all BOS points and EOL time) with some support for changing units (defaulting to days)
- `ResultsList.get_step_where` finds the index in `ResultsList` with a time closest to a requested point in time. Supports absolute and relative tolerances in the event you don't exactly know what point you seek.

Could be useful in conjunction with #1714 / #1726 to examine materials at a specific point.

Some follow up for discussions on this PR or future PRS

1. Supporting burnup units. 

Not immensely difficult since we have access to the necessary data (compositions, power). We could probably then add a method to the `Results` class just to compute the heavy metal loading at a given point in time and then make the conversion to MWd/kgHM based on the initial loading. 

2. Adding the `time_units` conversion into other methods (`get_eigenvalue` in particular)

We don't really need to use `get_times` to extract the time points in methods like `get_eigenvalue` because we grab those in the iteration. We may just want quick unit conversion (maybe a private method?) that the `ResultsList` can use to convert from the stored data in seconds to the desired units. 

Closes #1713

### CODE REVIEW

File: openmc/deplete/results_list.py [hunk 1]
```diff
@@ -192,3 +196,100 @@ def get_depletion_time(self):
         for ix, res in enumerate(items):
             times[ix] = res.proc_time
         return times
+
+    def get_times(self, time_units="d") -> np.ndarray:
```
Comment (paulromano) [inline openmc/deplete/results_list.py:200]: Hooray for type annotations!

File: openmc/deplete/results_list.py [hunk 2]
```diff
@@ -192,3 +196,100 @@ def get_depletion_time(self):
...
+
+        if time_units == "d":
+            times /= (60 * 60 * 24)
+        elif time_units == "h":
+            times /= (60 * 60)
+        elif time_units == "min":
+            times /= 60
+        elif time_units != "s":
+            raise ValueError(
+                f'Unable to set "time_units" to {time_units} since it is not '
```
Comment (paulromano) [inline openmc/deplete/results_list.py:231]: Right now, our stated support is [Python 3.5+](https://docs.openmc.org/en/latest/usersguide/install.html#usersguide-python-prereqs), although I'm in favor of moving to Python 3.6+ given that 3.5 is already past EOL and it's not even in our test matrix anymore (and man is it nice to have f-strings!). That being said I wanted to give others a chance to chime in here, especially @openmc-dev/technical-committee.

File: openmc/deplete/results_list.py [hunk 3]
```diff
@@ -192,3 +196,100 @@ def get_depletion_time(self):
...
+
+        Parameters
+        ----------
+        time : float
+            Desired point in time
+        time_units : {"s", "d", "min", "h"}, optional
+            Units on ``time``. Default: days
+        atol : float, optional
+            Absolute tolerance (in ``time_units``) if ``time`` is not
+            found. A value of ``None`` is converted to infinity.
```
Comment (paulromano) [inline openmc/deplete/results_list.py:254]: What do you think about just forcing users to pass `math.inf` if they really want this?

File: tests/unit_tests/test_deplete_resultslist.py [hunk 4]
```diff
@@ -65,3 +65,66 @@ def test_get_eigenvalue(res):
...
+
+    with pytest.raises(ValueError):
+        results.get_step_where(times[-1] + 1, time_units=unit, atol=0, rtol=0)
+
+    # Grab intermediate points with a small offset
+    delta = (times[1] - times[0])
+    offset = delta * 0.1
+    for expected, value in enumerate(times[1:-1], start=1):
+        # Shoot a little low and a little high
+        for mult in {1, -1}:
```
Comment (paulromano) [inline tests/unit_tests/test_deplete_resultslist.py:109]:
```suggestion
        for mult in (1, -1):
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the PR @drewejohnson! A few comments:

File: openmc/deplete/results_list.py (hunk 2 above)
Comment (drewejohnson) [reply openmc/deplete/results_list.py:231]: Oh yeah that totally slipped my mind. f-strings are just so nice! But not really worth dropping py 3.5 *just* for f-strings. I'm fine to change this back to fit with 3.5, but I do think a discussion with the technical committee would be fruitful.

Maybe for 0.13.0 release? I see 0.12.1 is listed as a milestone

File: openmc/deplete/results_list.py (hunk 3 above)
Comment (drewejohnson) [reply openmc/deplete/results_list.py:254]: Good with me! Very zen :v:

File: openmc/deplete/results_list.py (hunk 2 above)
Comment (drewejohnson) [reply openmc/deplete/results_list.py]: f-strings have been removed

File: openmc/deplete/results_list.py [hunk 5]
```diff
@@ -192,3 +196,97 @@ def get_depletion_time(self):
         for ix, res in enumerate(items):
             times[ix] = res.proc_time
         return times
+
+    def get_times(self, time_units="d") -> np.ndarray:
+        """Return the points in time that define the depletion schedule
+
+        Parameters
```
Comment (paulromano) [inline openmc/deplete/results_list.py:203]:
```suggestion

        .. versionadded:: 0.12.1

        Parameters
```

File: openmc/deplete/results_list.py [hunk 6]
```diff
@@ -192,3 +196,97 @@ def get_depletion_time(self):
...
+
+        In the event ``time`` lies exactly between two points, the
+        lower index will be returned. It is possible that the index
+        will be at most one past the point in time requested, but only
+        according to tolerances requested.
+
+        Passing ``atol=math.inf`` and ``rtol=math.inf`` will return
+        the closest index to the requested point.
+
+        Parameters
```
Comment (paulromano) [inline openmc/deplete/results_list.py:249]:
```suggestion

        .. versionadded:: 0.12.1

        Parameters
```

**Review by paulromano — CHANGES_REQUESTED** [review]
A few more suggested changes. Can you also merge `develop` into this branch to update the CI configuration?

**Review by paulromano — APPROVED** [review]
Thanks @drewejohnson!

**Pull request merged**